pkg/releaselock/controller.go

```diff
@@ -0,0 +1,203 @@
+		return nil, err
+	}
+
+	id, err := os.Hostname()
+	if err != nil {
+		klog.Errorf("Failed to get hostname for lockrelease controller: %v", err)
+		return nil, err
+	}
+	// Add a uniquifier so that two processes on the same host don't accidentally both become active.
+	id = id + "_" + string(uuid.NewUUID())
+
+	return &LockReleaseController{
+		id:                      id,
+		client:                  client,
+		leaderElection:          defaultLeaderElection,
+		leaderElectionNamespace: util.ConfigMapNamespace,
+		leaseDuration:           defaultLeaseDuration,
+		renewDeadline:           defaultRenewDeadline,
+		retryPeriod:             defaultRetryPeriod,
+		syncPeriod:              defaultSyncPeriod,
+	}, nil
+}
+
+func (c *LockReleaseController) Run(ctx context.Context) {
+	run := func(ctx context.Context) {
+		klog.Infof("Lock release controller %s started leading", c.id)
+		wait.Forever(func() {
+			cmList, err := c.client.CoreV1().ConfigMaps(util.ConfigMapNamespace).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				klog.Errorf("Failed to list configmap in namespace %s: %v", util.ConfigMapNamespace, err)
+			}
+			klog.Infof("Listed %d configmaps in namespace %s", len(cmList.Items), util.ConfigMapNamespace)
+			for _, cm := range cmList.Items {
+				// Filter out root ca.
+				if cm.Name == rootCA {
+					continue
+				}
+				if err := c.syncLockInfo(ctx, &cm); err != nil {
+					klog.Errorf("Failed to sync lock info for configmap %s/%s: %v", cm.Namespace, cm.Name, err)
+				}
+			}
+		}, defaultSyncPeriod)
+	}
+
+	rl, err := resourcelock.New(
+		resourcelock.LeasesResourceLock,
+		c.leaderElectionNamespace,
+		leaseName,
+		nil,
+		c.client.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity: c.id,
+		})
+	if err != nil {
+		klog.Fatalf("Error creating resourcelock: %v", err)
+	}
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: c.leaseDuration,
+		RenewDeadline: c.renewDeadline,
+		RetryPeriod:   c.retryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: run,
+			OnStoppedLeading: func() {
+				klog.Fatalf("%s no longer the leader, staying inactive.", c.id)
+			},
+		},
+	})
+}
+
+func (c *LockReleaseController) syncLockInfo(ctx context.Context, cm *corev1.ConfigMap) error {
+	nodeName, err := util.GKENodeNameFromConfigMap(cm)
+	if err != nil {
+		klog.Errorf("Failed to get GKE node name from configmap %s/%s: %v", cm.Namespace, cm.Name, err)
+		return err
+	}
+
+	klog.Infof("Getting GKE Node %s from API Server", nodeName)
+	node, err := c.client.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
+	if err != nil && !apiError.IsNotFound(err) {
+		klog.Errorf("Failed to get node %s: %v", nodeName, err)
+		return err
+	}
+
+	data := cm.DeepCopy().Data
+	recentCM := cm.DeepCopy()
+	for key, filestoreIP := range data {
+		_, _, _, _, gkeNodeID, gkeNodeInternalIP, err := util.ParseConfigMapKey(key)
```

Comment:
instead of gkeNodeID, gceInstanceID is a better name?

Reply:
done